src/Avalonia.Controls/RefreshVisualizer.cs

```diff
@@ -0,0 +1,226 @@
+            remove => RemoveHandler(RefreshRequestedEvent, value);
+        }
+
+        public RefreshVisualizer()
+        {
+            _visualizerRotateTransform = new RotateTransform();
+            _visualizerScaleTransform = new ScaleTransform()
+            {
+                ScaleX = 0.5f,
+                ScaleY = 0.5f,
+            };
+        }
+
+        protected override void OnApplyTemplate(TemplateAppliedEventArgs e)
+        {
+            base.OnApplyTemplate(e);
+
+            _visualizer = e.NameScope.Find<Visual>("PART_Visualizer");
+            _pivot = e.NameScope.Find<Visual>("PART_VisualizerPivot");
+
+            if (_visualizer != null)
+            {
+                RenderTransform = _visualizerScaleTransform;
+            }
+
+            if (_pivot != null)
+            {
+                _pivot.RenderTransform = _visualizerRotateTransform;
+            }
+        }
+
+        public void EndPull()
+        {
+            CheckAndSetRefreshing();
+        }
+
+        public void Pull(Vector delta)
+        {
+            if (_visualizer != null)
+            {
+                var yOffset = delta.Y * 0.4;
+
+                SetVisualizerTransforms(yOffset);
+                SetVisualizerState(yOffset);
+            }
+        }
+
+        private void SetVisualizerTransforms(double offset)
+        {
+            if (_visualizer != null && offset < DragLimit)
+            {
+                var ratio = offset / DragLimit;
+                var angle = ratio * 360;
+                _visualizerRotateTransform.Angle = angle;
+                Height = offset;
+                _visualizerScaleTransform.ScaleX = ratio < MinScale ? MinScale : ratio;
+                _visualizerScaleTransform.ScaleY = ratio < MinScale ? MinScale : ratio;
+                _visualizer.Opacity = Math.Min(ratio != 0 ? ratio + 0.3 : ratio, 1);
+            }
+        }
+
+        private void SetVisualizerState(double offset)
+        {
+            if (offset < DragLimit)
+            {
+
+                RefreshVisualizerState = RefreshVisualizerState.Interacting;
+            }
+            else
+            {
+                RefreshVisualizerState = RefreshVisualizerState.Pending;
+            }
+        }
+
+        private void CheckAndSetRefreshing()
+        {
+            if (RefreshVisualizerState == RefreshVisualizerState.Pending)
+            {
+                RefreshVisualizerState = RefreshVisualizerState.Refreshing;
+            }
+            else
+            {
+                RefreshVisualizerState = RefreshVisualizerState.Idle;
+            }
+        }
+    }
+
+    public enum RefreshVisualizerState
+    {
+        Idle,
+        Interacting,
+        Pending,
+        Refreshing
+    }
+
+    public class RefreshRequestedEventArgs : RoutedEventArgs
+    {
+        private RefreshCompletionDeferral _refreshCompletionDeferral;
+
+        public RefreshCompletionDeferral RefreshCompletionDeferral
```

Comment:
Incrementing ref-count in a property getter is counter-intuitive. Consider converting that into a method.

Reply:
> I see a lot of differences with the WinUI `RefreshContainer` and `RefreshVisualizer` controls. Do you know these controls well, are the differences on purpose? Generally, I would try to follow Microsoft unless there is a good reason to deviate. Personally, I don't use these controls so it would take a while for me to judge on a case-by-case basis good/bad API differences.
> 
> Edit: Ah, I realize you are probably following Maui here more than WinUI (at least that's where the RefreshView name comes from). Since these controls are mobile-focused that perhaps makes more sense. Wish Microsoft themselves was more consistent here though.

It now closely follows WinUI's implementation.